# ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆãƒ»ãƒ­ã‚¸ãƒƒã‚¯ä¾å­˜é–¢ä¿‚è¿½è·¡ã‚¬ã‚¤ãƒ‰

**å¯¾è±¡:** å…¨é–‹ç™ºè€…ãƒ»AI
**æœ€çµ‚æ›´æ–°:** 2025-10-22

---

## æ¦‚è¦

ã“ã®ã‚¬ã‚¤ãƒ‰ã¯ã€ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆãƒ»ãƒ­ã‚¸ãƒƒã‚¯é–“ã®ä¾å­˜é–¢ä¿‚ã‚’æ˜ç¢ºã«è¨˜éŒ²ãƒ»å¯è¦–åŒ–ã™ã‚‹æ–¹æ³•ã‚’å®šã‚ã¾ã™ã€‚

ãƒ•ã‚¡ã‚¤ãƒ«ã‚’é–‹ã„ãŸã¨ãã«ã€**è¦ªå­é–¢ä¿‚ãƒ»ä½¿ç”¨å…ˆãƒ»ä¾å­˜å…ˆãŒä¸€ç›®ç­ç„¶**ã«ãªã‚‹ã“ã¨ã§ã€ä»¥ä¸‹ã‚’å®Ÿç¾ï¼š

âœ… ä¿®æ­£ã®å½±éŸ¿ç¯„å›²ãŒå³åº§ã«åˆ¤å®šå¯èƒ½
âœ… ãƒªãƒ•ã‚¡ã‚¯ã‚¿ãƒªãƒ³ã‚°ã®ãƒªã‚¹ã‚¯è©•ä¾¡ãŒå®¹æ˜“
âœ… æ–°æ©Ÿèƒ½å®Ÿè£…æ™‚ã®è¨­è¨ˆåˆ¤æ–­ãŒæ˜ç¢º
âœ… ãƒ‡ãƒƒãƒ‰ã‚³ãƒ¼ãƒ‰æ¤œå‡ºãŒç°¡å˜

---

## æ–¹æ³• 1: ãƒ•ã‚¡ã‚¤ãƒ«ã‚³ãƒ¡ãƒ³ãƒˆï¼ˆè»½é‡ãƒ»æœ€æ¨å¥¨ï¼‰

### é…ç½®å ´æ‰€

å„ TypeScript / JavaScript ãƒ•ã‚¡ã‚¤ãƒ«ã® **å…ˆé ­ã‹ã‚‰ 5-10 è¡Œç›®**

### ãƒ†ãƒ³ãƒ—ãƒ¬ãƒ¼ãƒˆ

```typescript
/**
 * {Component/Logic Name}
 *
 * DEPENDENCY MAP:
 *
 * Parents (ã“ã®ãƒ•ã‚¡ã‚¤ãƒ«ã‚’ import ã—ã¦ã„ã‚‹å ´æ‰€):
 *   â”œâ”€ src/components/{ParentComponent}/index.tsx
 *   â”œâ”€ src/services/{Service}.ts
 *   â””â”€ src/pages/{Page}.tsx
 *
 * Dependencies (ã“ã®ãƒ•ã‚¡ã‚¤ãƒ«ãŒ import ã—ã¦ã„ã‚‹å¤–éƒ¨ãƒ•ã‚¡ã‚¤ãƒ«):
 *   â”œâ”€ src/hooks/useAuth.ts
 *   â”œâ”€ src/utils/validation.ts
 *   â””â”€ src/types/User.ts
 *
 * Related Files (é–¢é€£ãƒ•ã‚¡ã‚¤ãƒ«):
 *   â”œâ”€ Spec: ./Button.spec.md
 *   â”œâ”€ Tests: ./Button.test.tsx
 *   â”œâ”€ Styles: ./Button.module.css
 *   â””â”€ Issues: docs/issues/open/2025_10/20251022_01_button.md
 */
```

### å®Ÿè£…ä¾‹ 1: React ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ

```typescript
// src/components/Button/Button.tsx

/**
 * Button Component
 *
 * DEPENDENCY MAP:
 *
 * Parents (ä½¿ç”¨å…ˆ):
 *   â”œâ”€ src/components/Form/SubmitButton.tsx
 *   â”œâ”€ src/components/Modal/ConfirmDialog.tsx
 *   â”œâ”€ src/components/Navigation/NavBar.tsx
 *   â””â”€ src/pages/Dashboard.tsx
 *
 * Dependencies (ä¾å­˜å…ˆ):
 *   â”œâ”€ src/hooks/useClickHandler.ts
 *   â”œâ”€ src/utils/classNameBuilder.ts
 *   â””â”€ Button.module.css
 *
 * Related Files:
 *   â”œâ”€ Spec: ./Button.spec.md
 *   â”œâ”€ Tests: ./Button.test.tsx
 *   â””â”€ Styles: ./Button.module.css
 */

import React from 'react';
import { useClickHandler } from '@/hooks/useClickHandler';
import { classNameBuilder } from '@/utils/classNameBuilder';
import styles from './Button.module.css';

export interface ButtonProps {
  label: string;
  onClick: () => void;
  disabled?: boolean;
}

export const Button: React.FC<ButtonProps> = ({ label, onClick, disabled }) => {
  const handleClick = useClickHandler(onClick);
  const className = classNameBuilder(styles.button, { [styles.disabled]: disabled });

  return (
    <button className={className} onClick={handleClick} disabled={disabled}>
      {label}
    </button>
  );
};
```

### å®Ÿè£…ä¾‹ 2: ãƒ¦ãƒ¼ãƒ†ã‚£ãƒªãƒ†ã‚£é–¢æ•°

```typescript
// src/utils/calculateTotal.ts

/**
 * Calculate Total Price
 *
 * DEPENDENCY MAP:
 *
 * Parents (ä½¿ç”¨å…ˆ):
 *   â”œâ”€ src/components/Cart/CartSummary.tsx
 *   â”œâ”€ src/services/OrderService.ts
 *   â””â”€ src/hooks/useCart.ts
 *
 * Dependencies (ä¾å­˜å…ˆ):
 *   â”œâ”€ src/utils/applyDiscount.ts
 *   â”œâ”€ src/types/Item.ts
 *   â””â”€ src/constants/TAX_RATE.ts
 *
 * Related Files:
 *   â”œâ”€ Spec: ./calculateTotal.spec.md
 *   â”œâ”€ Tests: ./calculateTotal.test.ts
 *   â””â”€ Related: src/utils/applyDiscount.ts (same parent)
 */

import { applyDiscount } from "./applyDiscount";
import type { Item } from "@/types/Item";
import { TAX_RATE } from "@/constants/TAX_RATE";

export function calculateTotal(items: Item[], discountRate: number): number {
  const subtotal = items.reduce((sum, item) => sum + item.price * item.quantity, 0);
  const discounted = applyDiscount(subtotal, discountRate);
  return discounted * (1 + TAX_RATE);
}
```

### å®Ÿè£…ä¾‹ 3: Custom Hook

```typescript
// src/hooks/useAuth.ts

/**
 * useAuth Hook
 *
 * DEPENDENCY MAP:
 *
 * Parents (ä½¿ç”¨å…ˆ):
 *   â”œâ”€ src/components/Navigation/NavBar.tsx
 *   â”œâ”€ src/components/Auth/LoginForm.tsx
 *   â”œâ”€ src/pages/Dashboard.tsx
 *   â””â”€ src/components/ProtectedRoute.tsx
 *
 * Dependencies (ä¾å­˜å…ˆ):
 *   â”œâ”€ src/services/AuthService.ts
 *   â”œâ”€ src/context/AuthContext.ts
 *   â””â”€ src/types/User.ts
 *
 * Related Files:
 *   â”œâ”€ Spec: ./useAuth.spec.md
 *   â”œâ”€ Tests: ./useAuth.test.ts
 *   â””â”€ Context: src/context/AuthContext.ts (çŠ¶æ…‹ç®¡ç†)
 */

import { useContext, useCallback } from "react";
import { AuthContext } from "@/context/AuthContext";
import { AuthService } from "@/services/AuthService";
import type { User } from "@/types/User";

export function useAuth() {
  const { user, setUser } = useContext(AuthContext);
  const authService = new AuthService();

  const login = useCallback(
    async (email: string, password: string) => {
      const user = await authService.login(email, password);
      setUser(user);
      return user;
    },
    [setUser],
  );

  return { user, login };
}
```

---

## æ–¹æ³• 2: `.dependency.md` ãƒ•ã‚¡ã‚¤ãƒ«ï¼ˆè©³ç´°ç‰ˆï¼‰

### é…ç½®å ´æ‰€

å„ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ / ãƒ­ã‚¸ãƒƒã‚¯ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªã«é…ç½®ï¼š

```
src/components/Button/
â”œâ”€â”€ index.ts
â”œâ”€â”€ Button.tsx
â”œâ”€â”€ Button.spec.md           â† ä»•æ§˜
â”œâ”€â”€ Button.test.tsx          â† ãƒ†ã‚¹ãƒˆ
â”œâ”€â”€ Button.module.css        â† ã‚¹ã‚¿ã‚¤ãƒ«
â””â”€â”€ Button.dependency.md     â† ä¾å­˜é–¢ä¿‚è©³ç´°ï¼ˆNEWï¼‰
```

### ãƒ†ãƒ³ãƒ—ãƒ¬ãƒ¼ãƒˆ

```markdown
# Button.dependency.md

## ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆæƒ…å ±

- **ãƒ•ã‚¡ã‚¤ãƒ«ãƒ‘ã‚¹**: `src/components/Button/Button.tsx`
- **ä½œæˆæ—¥**: 2025-10-22
- **æœ€çµ‚æ›´æ–°**: 2025-10-22
- **ã‚«ãƒ†ã‚´ãƒª**: UI Component

---

## ä¾å­˜é–¢ä¿‚ãƒãƒƒãƒ—

### è¦ªã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ / ä½¿ç”¨å…ˆ (Dependents)

ã“ã®ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆã‚’ import ã—ã¦ã„ã‚‹å ´æ‰€ï¼ˆ**ä¿®æ­£æ™‚ã«å½±éŸ¿ã‚’å—ã‘ã‚‹å…ˆ**ï¼‰ï¼š

| ä½¿ç”¨å…ˆ        | ãƒ•ã‚¡ã‚¤ãƒ«ãƒ‘ã‚¹                           | ç”¨é€”                         | ãƒ†ã‚¹ãƒˆçŠ¶æ³ |
| ------------- | -------------------------------------- | ---------------------------- | ---------- |
| SubmitButton  | src/components/Form/SubmitButton.tsx   | ãƒ•ã‚©ãƒ¼ãƒ é€ä¿¡ãƒœã‚¿ãƒ³ã®ãƒ©ãƒƒãƒ‘ãƒ¼ | âœ… Tested  |
| ConfirmDialog | src/components/Modal/ConfirmDialog.tsx | ç¢ºèªãƒ€ã‚¤ã‚¢ãƒ­ã‚°ã®ãƒœã‚¿ãƒ³       | âœ… Tested  |
| NavBar        | src/components/Navigation/NavBar.tsx   | ãƒŠãƒ“ã‚²ãƒ¼ã‚·ãƒ§ãƒ³ãƒãƒ¼ã®ãƒœã‚¿ãƒ³   | âœ… Tested  |
| Dashboard     | src/pages/Dashboard.tsx                | ãƒšãƒ¼ã‚¸å†…ã®ã‚¢ã‚¯ã‚·ãƒ§ãƒ³         | âš ï¸ Partial |

**ãƒªã‚¹ã‚¯è©•ä¾¡**: 4 ã‹æ‰€ã§ä½¿ç”¨ â†’ **ä¿®æ­£æ™‚ã®å½±éŸ¿ç¯„å›²ã¯å¤§ãã„**

---

### ä¾å­˜ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ / ã‚¤ãƒ³ãƒãƒ¼ãƒˆå…ˆ (Dependencies)

ã“ã®ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆãŒ import ã—ã¦ã„ã‚‹å¤–éƒ¨ãƒ•ã‚¡ã‚¤ãƒ«ï¼ˆ**ä¿®æ­£æ™‚ã«ä¾å­˜ã™ã‚‹å…ˆ**ï¼‰ï¼š

| ä¾å­˜å…ˆ            | ãƒ•ã‚¡ã‚¤ãƒ«ãƒ‘ã‚¹                  | ç”¨é€”                 | å¿…é ˆåº¦      |
| ----------------- | ----------------------------- | -------------------- | ----------- |
| useClickHandler   | src/hooks/useClickHandler.ts  | ã‚¯ãƒªãƒƒã‚¯ãƒãƒ³ãƒ‰ãƒªãƒ³ã‚° | ğŸ”´ Critical |
| classNameBuilder  | src/utils/classNameBuilder.ts | ã‚¯ãƒ©ã‚¹åç”Ÿæˆ         | ğŸŸ¡ High     |
| Button.module.css | ./Button.module.css           | ã‚¹ã‚¿ã‚¤ãƒªãƒ³ã‚°         | ğŸŸ¡ High     |

**ãƒªã‚¹ã‚¯è©•ä¾¡**: useClickHandler ã«ä¾å­˜ â†’ **Hook ã®ä¿®æ­£æ™‚ã«å½±éŸ¿ã‚’å—ã‘ã‚‹**

---

## é–¢é€£ãƒ•ã‚¡ã‚¤ãƒ«

### åŒã˜æ©Ÿèƒ½ã«é–¢é€£
```

src/components/Button/
â”œâ”€â”€ Button.tsx â† Main component
â”œâ”€â”€ Button.spec.md â† ä»•æ§˜å®šç¾©
â”œâ”€â”€ Button.test.tsx â† ãƒ†ã‚¹ãƒˆã‚³ãƒ¼ãƒ‰
â”œâ”€â”€ Button.module.css â† ã‚¹ã‚¿ã‚¤ãƒ«
â””â”€â”€ Button.dependency.md â† ã“ã®ãƒ•ã‚¡ã‚¤ãƒ«

```

### ä»•æ§˜ãƒ»ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆ

- **ä»•æ§˜æ›¸**: `Button.spec.md`
- **å®Ÿè£…è¨ˆç”»**: `docs/03_plans/button-component/20251022_01_implementation-plan.md`
- **é–¢é€£ Issue**: `docs/01_issues/open/2025_10/20251022_01_button-component.md`

---

## ä¾å­˜ã‚°ãƒ©ãƒ•

```

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ Button Component â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ â”‚
â”‚ INCOMING (ã“ã®ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆã‚’ä½¿ç”¨): â”‚
â”‚ â”Œâ”€ SubmitButton â”€â” â”‚
â”‚ â”œâ”€ ConfirmDialog â”¤ â”‚
â”‚ â”œâ”€ NavBar â”€â”€â”€â”€â”€â”€â”€â”¤ â”‚
â”‚ â””â”€ Dashboard â”€â”€â”€â”€â”˜ â”‚
â”‚ â”‚ â”‚
â”‚ â”‚ (ä¿®æ­£æ™‚ã®å½±éŸ¿ç¯„å›²) â”‚
â”‚ â–¼ â”‚
â”‚ â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â” â”‚
â”‚ â”‚ Button.tsx â”‚ â”‚
â”‚ â”‚ (ã“ã®ãƒ•ã‚¡ã‚¤ãƒ«) â”‚ â”‚
â”‚ â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜ â”‚
â”‚ â–² â”‚
â”‚ â”‚ (ä¿®æ­£æ™‚ã«ä¾å­˜) â”‚
â”‚ â”‚ â”‚
â”‚ OUTGOING (ã“ã®ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆãŒä½¿ç”¨): â”‚
â”‚ â”Œâ”€ useClickHandler â” â”‚
â”‚ â”œâ”€ classNameBuilderâ”¤ â”‚
â”‚ â””â”€ Button.css â”€â”€â”€â”€â”˜ â”‚
â”‚ â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

````

---

## ä½¿ç”¨å ´é¢åˆ¥ã‚¬ã‚¤ãƒ‰

### ä¿®æ­£ã™ã‚‹ã¨ã

ã“ã®ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆã‚’ä¿®æ­£ã™ã‚‹éš›ã€**è¦ªã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ** ã®åˆ—ã‚’ç¢ºèªï¼š

> âœ… **ãƒã‚§ãƒƒã‚¯é …ç›®**:
> 1. SubmitButton ã§ãƒ†ã‚¹ãƒˆãŒé€šã‚‹ã‹
> 2. ConfirmDialog ã§ãƒ†ã‚¹ãƒˆãŒé€šã‚‹ã‹
> 3. NavBar ã§ãƒ†ã‚¹ãƒˆãŒé€šã‚‹ã‹
> 4. Dashboard ã§ãƒ†ã‚¹ãƒˆãŒé€šã‚‹ã‹

**å¯¾å¿œ**: ä¿®æ­£å¾Œã€è¦ªã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆã®ãƒ†ã‚¹ãƒˆã‚’å…¨ã¦å®Ÿè¡Œ

---

### å‰Šé™¤ã™ã‚‹ã¨ã

å‰Šé™¤å‰ã«è¦ªã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ ã‚’ç¢ºèªã—ã¦ã€ä¾å­˜é–¢ä¿‚ã‚’è§£æ±ºï¼š

> âš ï¸ **å‰Šé™¤ä¸å¯**: 4 ã‹æ‰€ã§ä½¿ç”¨ä¸­
>
> å¯¾å¿œ:
> 1. ä½¿ç”¨å…ˆ 4 ã‹æ‰€ã‚’æ–°ã—ã„å®Ÿè£…ã«ç½®ãæ›ãˆ
> 2. ãƒ†ã‚¹ãƒˆã‚’å…¨ã¦ pass ã•ã›ã‚‹
> 3. ãã®å¾Œã«å‰Šé™¤

---

### ãƒªãƒ•ã‚¡ã‚¯ã‚¿ãƒªãƒ³ã‚°

ä¾å­˜é–¢ä¿‚ã‚’åˆ†æã—ã¦ãƒªãƒ•ã‚¡ã‚¯ã‚¿ãƒªãƒ³ã‚°è¨ˆç”»ã‚’ç«‹æ¡ˆï¼š

> **ç¾åœ¨ã®æ§‹é€ **:
> - Button â†’ useClickHandler â†’ AuthService
>
> **ãƒªã‚¹ã‚¯**: useClickHandler ä¿®æ­£æ™‚ã« Button ã‚‚å½±éŸ¿
>
> **å¯¾å¿œ**: ä¾å­˜é–¢ä¿‚ã‚’ç·©å’Œã™ã‚‹è¨­è¨ˆå¤‰æ›´ã‚’æ¤œè¨

---

## æ›´æ–°ãƒ«ãƒ¼ãƒ«

### ã„ã¤æ›´æ–°ã™ã‚‹ã‹

âœ… **å¿…ãšæ›´æ–°**:
- æ–°ã—ã„è¦ªã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ ãŒä½¿ç”¨ã‚’é–‹å§‹
- ä¾å­˜ãƒ•ã‚¡ã‚¤ãƒ«ã‚’è¿½åŠ ãƒ»å‰Šé™¤
- å¤§è¦æ¨¡ãƒªãƒ•ã‚¡ã‚¯ã‚¿ãƒªãƒ³ã‚°

âŒ **æ›´æ–°ä¸è¦**:
- ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆå†…ã®ãƒ­ã‚¸ãƒƒã‚¯ã®ã¿ä¿®æ­£
- CSS ã®ç´°ã‹ã„èª¿æ•´

### æ›´æ–°æ‰‹é †

1. ãƒ•ã‚¡ã‚¤ãƒ«å…ˆé ­ã® DEPENDENCY MAP ã‚³ãƒ¡ãƒ³ãƒˆã‚’æ›´æ–°
2. `.dependency.md` ã®ãƒ†ãƒ¼ãƒ–ãƒ«ã‚‚æ›´æ–°
3. é–¢é€£ã™ã‚‹ä»–ã® `.dependency.md` ã‚‚ç¢ºèªãƒ»æ›´æ–°

---

## è‡ªå‹•åŒ–ãƒ„ãƒ¼ãƒ«

### Tool 1: madge (ä¾å­˜ã‚°ãƒ©ãƒ•è‡ªå‹•ç”Ÿæˆ)

```bash
# ã‚¤ãƒ³ã‚¹ãƒˆãƒ¼ãƒ«
npm install --save-dev madge

# ã‚°ãƒ©ãƒ•ç”Ÿæˆ
npx madge --extensions ts,tsx src/components/Button/Button.tsx

# ãƒ“ã‚¸ãƒ¥ã‚¢ãƒ«å‡ºåŠ›
npx madge --image deps.svg src/
````

### Tool 2: ãƒ‡ãƒãƒƒã‚°ã‚¹ã‚¯ãƒªãƒ—ãƒˆ

```bash
# è¦ªã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆæ¤œç´¢
grep -r "from.*Button" src/ --include="*.tsx"

# ä¾å­˜å…ˆæ¤œç´¢
grep "^import\|^from" src/components/Button/Button.tsx
```

### Tool 3: VS Code æ‹¡å¼µæ©Ÿèƒ½

æ¨å¥¨:

- **Dependency Cruiser**: ä¾å­˜é–¢ä¿‚ãƒ“ã‚¸ãƒ¥ã‚¢ãƒ©ã‚¤ã‚¶ãƒ¼
- **Import Cost**: ã‚¤ãƒ³ãƒãƒ¼ãƒˆå½±éŸ¿åº¦è¡¨ç¤º

---

## ãƒ™ã‚¹ãƒˆãƒ—ãƒ©ã‚¯ãƒ†ã‚£ã‚¹

### âœ… Good

```typescript
/**
 * Button Component
 *
 * DEPENDENCY MAP:
 *
 * Parents (ä½¿ç”¨å…ˆ):
 *   â”œâ”€ SubmitButton
 *   â””â”€ NavBar
 *
 * Dependencies (ä¾å­˜å…ˆ):
 *   â”œâ”€ useClickHandler
 *   â””â”€ styles.css
 */
```

### âŒ Bad

```typescript
// No dependency information
// ä½•ã«ã‚‚ä½¿ã‚ã‚Œã¦ãªã„ï¼Ÿã‚ã‹ã‚‰ãªã„

export const Button = () => {};
```

---

## ãƒã‚§ãƒƒã‚¯ãƒªã‚¹ãƒˆ

æ–°è¦ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆä½œæˆæ™‚ï¼š

- [ ] ãƒ•ã‚¡ã‚¤ãƒ«å…ˆé ­ã« DEPENDENCY MAP ã‚³ãƒ¡ãƒ³ãƒˆãŒã‚ã‚‹ã‹
- [ ] Parents (ä½¿ç”¨å…ˆ) ã‚’è¨˜è¼‰ã—ãŸã‹ï¼ˆä½œæˆæ™‚ã¯ç©ºã§ã‚‚ OKï¼‰
- [ ] Dependencies (ä¾å­˜å…ˆ) ã‚’è¨˜è¼‰ã—ãŸã‹
- [ ] Related Files ã‚’è¨˜è¼‰ã—ãŸã‹
- [ ] `.dependency.md` ã‚’ä½œæˆã—ãŸã‹ï¼ˆå¤§è¦æ¨¡ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆã®å ´åˆï¼‰

ä¿®æ­£ãƒ»ãƒªãƒ•ã‚¡ã‚¯ã‚¿ãƒªãƒ³ã‚°æ™‚ï¼š

- [ ] DEPENDENCY MAP ã‚’æ›´æ–°ã—ãŸã‹
- [ ] è¦ªã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆã®ãƒ†ã‚¹ãƒˆã‚’å®Ÿè¡Œã—ãŸã‹
- [ ] ä¾å­˜å…ˆã®å¤‰æ›´ãŒãªã„ã‹ç¢ºèªã—ãŸã‹

---

## å®Ÿè£…ä¾‹ä¸€è¦§

- [Button Component](#å®Ÿè£…ä¾‹-1-react-ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ) (React ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ)
- [calculateTotal](#å®Ÿè£…ä¾‹-2-ãƒ¦ãƒ¼ãƒ†ã‚£ãƒªãƒ†ã‚£é–¢æ•°) (ãƒ¦ãƒ¼ãƒ†ã‚£ãƒªãƒ†ã‚£)
- [useAuth](#å®Ÿè£…ä¾‹-3-custom-hook) (Custom Hook)

---

## è‡ªå‹•åŒ–ãƒ„ãƒ¼ãƒ«

### ä¾å­˜é–¢ä¿‚ã‚’è‡ªå‹•çš„ã«å¯è¦–åŒ–

è»½é‡ç‰ˆï¼ˆãƒ•ã‚¡ã‚¤ãƒ«ã‚³ãƒ¡ãƒ³ãƒˆï¼‰ã§æ—¥ã€…ã®é–‹ç™ºã‚’å®Ÿæ–½ã—ã¤ã¤ã€å®šæœŸçš„ã«è‡ªå‹•ãƒ„ãƒ¼ãƒ«ã§ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆå…¨ä½“ã®ä¾å­˜é–¢ä¿‚ã‚’åˆ†æãƒ»å¯è¦–åŒ–ã—ã¾ã™ã€‚

#### æ¨å¥¨ãƒ„ãƒ¼ãƒ«: **Madge**

```bash
# ã‚¤ãƒ³ã‚¹ãƒˆãƒ¼ãƒ«
npm install --save-dev madge

# å¾ªç’°ä¾å­˜ã‚’ãƒã‚§ãƒƒã‚¯ï¼ˆé–‹ç™ºæ™‚ï¼‰
npx madge src/ --circular

# ãƒ€ã‚¤ã‚¢ã‚°ãƒ©ãƒ ã‚’ç”Ÿæˆï¼ˆå®šæœŸçš„ï¼‰
npx madge src/ --image dist/dependencies.svg

# JSON å‡ºåŠ›ï¼ˆãƒ—ãƒ­ã‚°ãƒ©ãƒ å‡¦ç†ç”¨ï¼‰
npx madge src/ --json > deps.json
```

#### ãã®ä»–ã®ãƒ„ãƒ¼ãƒ«

| ãƒ„ãƒ¼ãƒ«                 | ç”¨é€”                             | æ¨å¥¨åº¦     |
| ---------------------- | -------------------------------- | ---------- |
| **Madge**              | ä¾å­˜ã‚°ãƒ©ãƒ•ã®å¯è¦–åŒ–ãƒ»å¾ªç’°ä¾å­˜æ¤œå‡º | â­â­â­â­â­ |
| **Dependency Cruiser** | è©³ç´°ãƒ«ãƒ¼ãƒ«æ¤œè¨¼ãƒ»é«˜åº¦ãªåˆ†æ       | â­â­â­â­   |
| **depcheck**           | æœªä½¿ç”¨ä¾å­˜é–¢ä¿‚ã®æ¤œå‡º             | â­â­â­     |
| **Graphviz**           | æ‰‹å‹•ãƒ€ã‚¤ã‚¢ã‚°ãƒ©ãƒ ç”Ÿæˆ             | â­â­â­â­   |

è©³ç´°ã¯ [ä¾å­˜é–¢ä¿‚ãƒ€ã‚¤ã‚¢ã‚°ãƒ©ãƒ è‡ªå‹•ç”Ÿæˆãƒ„ãƒ¼ãƒ«](./dependency-visualization-tools.md) ã‚’å‚ç…§ã€‚

### npm scripts ä¾‹

```json
{
  "scripts": {
    "analyze:deps": "madge src/",
    "analyze:circular": "madge src/ --circular",
    "visualize:deps": "madge src/ --image dist/dependencies.svg",
    "check:deps": "npm run analyze:circular -- --exit 1"
  }
}
```

### CI ãƒ‘ã‚¤ãƒ—ãƒ©ã‚¤ãƒ³çµ±åˆ

```yaml
# .github/workflows/dependency-check.yml
- name: Check circular dependencies
  run: npx madge src/ --circular --exit 1

- name: Generate dependency graph
  run: npx madge src/ --image graph.svg
```

---

## ğŸ”— é–¢é€£ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆ

- [å‘½åè¦å‰‡](./naming-conventions.md)
- [AI ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆé§†å‹•é–‹ç™ºã‚¬ã‚¤ãƒ‰](./ai-documentation.md)
- [ã‚³ãƒ¼ãƒ‰å“è³ªåŸºæº–](./code-quality-standards.md)
- [ä¾å­˜é–¢ä¿‚ãƒ€ã‚¤ã‚¢ã‚°ãƒ©ãƒ è‡ªå‹•ç”Ÿæˆãƒ„ãƒ¼ãƒ«](./dependency-visualization-tools.md) â† è©³ç´°ãªä½¿ç”¨æ–¹æ³•

---

**æœ€çµ‚æ›´æ–°:** 2025-10-22
**ä½œæˆè€…:** AI (Grok Code Fast 1)
