# è¨€èªè¦å‰‡

**å¯¾è±¡:** å…¨é–‹ç™ºè€…
**æœ€çµ‚æ›´æ–°:** 2025-10-22

---

## æ¦‚è¦

ã“ã®ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã§ã¯ã€ã‚³ãƒ¼ãƒ‰ã®ç”¨é€”ã«å¿œã˜ã¦è¨€èªã‚’ä½¿ã„åˆ†ã‘ã‚‹ã“ã¨ã§ã€ä»¥ä¸‹ã‚’å®Ÿç¾ã—ã¾ã™ï¼š

- ã‚°ãƒ­ãƒ¼ãƒãƒ«ãªé–‹ç™ºç’°å¢ƒã§ã®å¯èª­æ€§ã‚’ç¢ºä¿
- æ—¥æœ¬èªè©±è€…ãƒãƒ¼ãƒ ã¨ã® ã‚³ãƒŸãƒ¥ãƒ‹ã‚±ãƒ¼ã‚·ãƒ§ãƒ³åŠ¹ç‡åŒ–
- AI ãƒ¢ãƒ‡ãƒ«ã¨ã®é€£æºç²¾åº¦å‘ä¸Š

---

## 1. ã‚³ãƒ¼ãƒ‰å†…ã‚³ãƒ¡ãƒ³ãƒˆï¼ˆè‹±èªï¼‰

### 1.1 åŸºæœ¬ãƒ«ãƒ¼ãƒ«

```typescript
// Use English for code comments
// è‹±èªã§ã‚³ãƒ¡ãƒ³ãƒˆã‚’è¨˜è¿°ã—ã¦ãã ã•ã„

// âœ… Good
function calculateTotal(items: Item[]): number {
  // Filter out items with zero price
  const validItems = items.filter((item) => item.price > 0);

  // Sum all item prices
  return validItems.reduce((sum, item) => sum + item.price, 0);
}

// âŒ Bad
function calculateTotal(items: Item[]): number {
  // ã‚¼ãƒ­ä¾¡æ ¼ã®å•†å“ã‚’ãƒ•ã‚£ãƒ«ã‚¿ãƒªãƒ³ã‚°
  const validItems = items.filter((item) => item.price > 0);

  // ã™ã¹ã¦ã®å•†å“ã®ä¾¡æ ¼ã‚’åˆè¨ˆ
  return validItems.reduce((sum, item) => sum + item.price, 0);
}
```

### 1.2 ä½•ã‚’ã€ãªãœã€ã©ã†ã™ã‚‹ã‹

ã‚³ãƒ¡ãƒ³ãƒˆã¯ã€Œãªãœã€ã‚’èª¬æ˜ã—ã¾ã™ã€‚ã€Œä½•ã‚’ã€ã¯ã‚³ãƒ¼ãƒ‰ã§è¡¨ç¾ã—ã¦ãã ã•ã„ã€‚

```typescript
// âŒ Bad: ä½•ã‚’ã—ã¦ã„ã‚‹ã‹èª¬æ˜ï¼ˆã‚³ãƒ¼ãƒ‰ãŒæ˜ç¢ºã ã‹ã‚‰ä¸è¦ï¼‰
let x = arr.length; // Get array length

// âœ… Good: ãªãœå¿…è¦ã‹èª¬æ˜
let maxRetries = 3;  // Limit retries to prevent infinite loops

// âŒ Bad: é•·ã™ãã‚‹ã‚³ãƒ¡ãƒ³ãƒˆ
// This function processes user input by validating email format
// and checking if the user exists in the database
function validateUser(email: string): boolean {

// âœ… Good: ç°¡æ½”ã§ç›®çš„ã‚’æ˜ç¢ºã«
function validateUser(email: string): boolean {
  // Return false if email format is invalid
```

### 1.3 ã‚³ãƒ¡ãƒ³ãƒˆã®ç¨®é¡ã¨å½¢å¼

#### é–¢æ•°ãƒ»ã‚¯ãƒ©ã‚¹ã®èª¬æ˜ï¼ˆJSDocï¼‰

```typescript
/**
 * Calculates total price of items after applying discounts.
 *
 * @param items - Array of items to process
 * @param discountRate - Discount rate (0-1)
 * @returns Total price after discount
 * @throws Error if items array is empty
 * @example
 * const total = calculateDiscount([item1, item2], 0.1);
 */
function calculateDiscount(items: Item[], discountRate: number): number {
  if (items.length === 0) {
    throw new Error("Items array must not be empty");
  }

  const subtotal = items.reduce((sum, item) => sum + item.price, 0);
  return subtotal * (1 - discountRate);
}
```

#### è¤‡é›‘ãªãƒ­ã‚¸ãƒƒã‚¯ã®èª¬æ˜

```typescript
// Algorithm explanation when needed
function fibonacci(n: number): number {
  // Base cases: fib(0)=0, fib(1)=1
  if (n <= 1) return n;

  // Use iterative approach instead of recursion to avoid stack overflow
  // Time complexity: O(n), Space complexity: O(1)
  let prev = 0,
    curr = 1;
  for (let i = 2; i <= n; i++) {
    [prev, curr] = [curr, prev + curr];
  }
  return curr;
}
```

#### TODOãƒ»FIXME ã‚³ãƒ¡ãƒ³ãƒˆ

```typescript
// TODO: Optimize this function to handle large datasets
function processLargeData(data: any[]): any[] {
  // ...
}

// FIXME: Race condition detected when concurrent requests
async function fetchUserData(userId: number): Promise<User> {
  // ...
}

// NOTE: This workaround is needed due to browser compatibility issue
// See: https://github.com/issue/12345
function handleSpecialKeyEvent(e: KeyboardEvent): void {
  // ...
}
```

### 1.4 é¿ã‘ã‚‹ã¹ãã‚³ãƒ¡ãƒ³ãƒˆ

```typescript
// âŒ é¿ã‘ã‚‹: æ—¥æœ¬èªã‚³ãƒ¡ãƒ³ãƒˆ
const userName = "John"; // ãƒ¦ãƒ¼ã‚¶ãƒ¼å

// âŒ é¿ã‘ã‚‹: éåº¦ã«è©³ç´°
const x = 5; // Set x to 5

// âŒ é¿ã‘ã‚‹: ã‚³ãƒ¼ãƒ‰ãŒæ˜ç¢ºãªã‚‰ä¸è¦
const isValid = email.includes("@"); // Check if email contains @

// âŒ é¿ã‘ã‚‹: å¤ã„/å»ƒæ­¢äºˆå®šã‚³ãƒ¼ãƒ‰
// const oldFunction = () => { ... };  // Don't use this

// âœ… ä»£æ›¿æ¡ˆ: æ˜ç¢ºãª TODO ã‹å‰Šé™¤
// TODO: Remove this function in v2.0
// function oldFunction() { ... }
```

---

## 2. ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆï¼ˆæ—¥æœ¬èªï¼‰

### 2.1 å¯¾è±¡ãƒ•ã‚¡ã‚¤ãƒ«

ä»¥ä¸‹ã®ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆã¯æ—¥æœ¬èªã§è¨˜è¿°ï¼š

```
docs/
â”œâ”€â”€ README.md                        # æ—¥æœ¬èª
â”œâ”€â”€ 01_planning/
â”‚   â””â”€â”€ *.md                         # æ—¥æœ¬èª
â”œâ”€â”€ 02_requirements/
â”‚   â””â”€â”€ *.md                         # æ—¥æœ¬èª
â”œâ”€â”€ 03_design/
â”‚   â””â”€â”€ *.md                         # æ—¥æœ¬èª
â”œâ”€â”€ 04_implementation/
â”‚   â””â”€â”€ *.md                         # æ—¥æœ¬èª
â”œâ”€â”€ 07_research/
â”‚   â””â”€â”€ *.md                         # æ—¥æœ¬èª
â”œâ”€â”€ 08_worklogs/
â”‚   â””â”€â”€ *.md                         # æ—¥æœ¬èª
â”œâ”€â”€ issues/
â”‚   â””â”€â”€ *.md                         # æ—¥æœ¬èª
â””â”€â”€ rules/
    â””â”€â”€ *.md                         # æ—¥æœ¬èª
```

### 2.2 ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆä½œæˆãƒ«ãƒ¼ãƒ«

````markdown
# æ©Ÿèƒ½ã‚¿ã‚¤ãƒˆãƒ«ï¼ˆæ—¥æœ¬èªï¼‰

**å¯¾è±¡:** {å¯¾è±¡è€…}
**æœ€çµ‚æ›´æ–°:** YYYY-MM-DD

## ã‚»ã‚¯ã‚·ãƒ§ãƒ³1: æ¦‚è¦

å†…å®¹ã‚’æ—¥æœ¬èªã§è¨˜è¿°ã—ã¾ã™ã€‚

### å°ã‚»ã‚¯ã‚·ãƒ§ãƒ³

è©³ç´°ã‚’æ—¥æœ¬èªã§è¨˜è¿°ã—ã¾ã™ã€‚

## ã‚»ã‚¯ã‚·ãƒ§ãƒ³2: å®Ÿè£…ä¾‹

ã‚³ãƒ¼ãƒ‰ä¾‹ï¼š

```typescript
// Code snippet
const result = calculate();
```
````

## é–¢é€£ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆ

- [ãƒ•ã‚¡ã‚¤ãƒ«å](./path/to/file.md)
- [å¤–éƒ¨ãƒªãƒ³ã‚¯](https://example.com)

---

**æœ€çµ‚æ›´æ–°:** YYYY-MM-DD

````

### 2.3 æŠ€è¡“ç”¨èªã®è¡¨è¨˜

```markdown
âœ… æ¨å¥¨: è‹±èªæŠ€è¡“ç”¨èªã¯ãã®ã¾ã¾ä½¿ç”¨
- React ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ
- API ãƒ¬ã‚¹ãƒãƒ³ã‚¹
- Jest ãƒ†ã‚¹ãƒˆ
- TypeScript å‹å®šç¾©

âœ… æ—¥æœ¬èªåŒ–ã™ã‚‹å ´åˆã¯çµ±ä¸€
- ä¸€åº¦å®šç¾©ã—ãŸã‚‰ã€ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆå…¨ä½“ã§åŒã˜è¡¨è¨˜ã‚’ä½¿ç”¨

âŒ æ··åœ¨ã•ã›ãªã„: åŒã˜ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆå†…ã§ã®ä¸çµ±ä¸€
- API ãƒ¬ã‚¹ãƒãƒ³ã‚¹ vs API å¿œç­”ï¼ˆæ··åœ¨ã•ã›ãªã„ï¼‰
````

### 2.4 ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆä¾‹

````markdown
# Button ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆå®Ÿè£…è¨ˆç”»

## æ¦‚è¦

Button ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆã¯ã€ãƒ¦ãƒ¼ã‚¶ãƒ¼ã‚¤ãƒ³ã‚¿ãƒ©ã‚¯ã‚·ãƒ§ãƒ³ã®åŸºæœ¬å˜ä½ã§ã™ã€‚
è¤‡æ•°ã®ã‚µã‚¤ã‚ºã¨ãƒãƒªã‚¢ãƒ³ãƒˆã‚’ã‚µãƒãƒ¼ãƒˆã—ã¾ã™ã€‚

## ãƒ•ã‚§ãƒ¼ã‚º1: åŸºæœ¬æ©Ÿèƒ½

- ãƒ†ã‚­ã‚¹ãƒˆãƒ¬ãƒ³ãƒ€ãƒªãƒ³ã‚°
- onClick ã‚¤ãƒ™ãƒ³ãƒˆãƒãƒ³ãƒ‰ãƒªãƒ³ã‚°
- disabled çŠ¶æ…‹

## å®Ÿè£…ä¾‹

src/components/Button/Button.tsx:

```typescript
export const Button: React.FC<ButtonProps> = (props) => {
  return <button {...props}>{props.children}</button>;
};
```
````

````

---

## 3. é–¢æ•°åãƒ»å¤‰æ•°åï¼ˆè‹±èªï¼‰

### 3.1 åŸºæœ¬ãƒ«ãƒ¼ãƒ«

```typescript
// âœ… Good: è‹±èª
const userName = 'John';
const calculateTotal = () => { };
class UserManager { }
interface ButtonProps { }

// âŒ Bad: æ—¥æœ¬èª
const ãƒ¦ãƒ¼ã‚¶ãƒ¼å = 'John';
const åˆè¨ˆã‚’è¨ˆç®— = () => { };
class ãƒ¦ãƒ¼ã‚¶ãƒ¼ç®¡ç† { }
interface ãƒœã‚¿ãƒ³ãƒ—ãƒ­ãƒƒãƒ— { }
````

### 3.2 ç†ç”±

1. **IDE ã‚µãƒãƒ¼ãƒˆ**: è‹±èªã®æ–¹ãŒ autocomplete ãŒå„ªã‚Œã¦ã„ã‚‹
2. **å›½éš›çš„ãªã‚³ãƒ©ãƒœãƒ¬ãƒ¼ã‚·ãƒ§ãƒ³**: è‹±èªãŒæ¨™æº–
3. **ãƒ©ã‚¤ãƒ–ãƒ©ãƒªäº’æ›æ€§**: npm ãƒ‘ãƒƒã‚±ãƒ¼ã‚¸ã¯è‹±èªå‘½åãŒæ…£ä¾‹
4. **æ¤œç´¢æ€§**: è‹±èªã‚­ãƒ¼ãƒ¯ãƒ¼ãƒ‰ã§æ¤œç´¢ã—ã‚„ã™ã„

---

## 4. ã‚³ãƒŸãƒƒãƒˆãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ï¼ˆè‹±èªï¼‰

### 4.1 Conventional Commits å½¢å¼

```
<type>(<scope>): <subject>
<BLANK LINE>
<body>
<BLANK LINE>
<footer>
```

### 4.2 ä¾‹

```
feat(Button): Add loading state support

Add isLoading prop to Button component to display loading spinner
during async operations.

- Add loading animation CSS
- Add TC-006 test case
- Update Button.spec.md

Fixes #123
Related to: docs/issues/open/2025_10/20251022_01_button-component.md
```

### 4.3 Type ä¸€è¦§

| Type     | èª¬æ˜               | ä¾‹                                           |
| -------- | ------------------ | -------------------------------------------- |
| feat     | æ–°æ©Ÿèƒ½è¿½åŠ          | feat(Button): Add size variant               |
| fix      | ãƒã‚°ä¿®æ­£           | fix(Modal): Fix z-index issue                |
| docs     | ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆä¿®æ­£   | docs: Update naming conventions              |
| test     | ãƒ†ã‚¹ãƒˆè¿½åŠ ãƒ»ä¿®æ­£   | test(Button): Add TC-005                     |
| refactor | ãƒªãƒ•ã‚¡ã‚¯ã‚¿ãƒªãƒ³ã‚°   | refactor(utils): Simplify calculate function |
| chore    | ãƒ“ãƒ«ãƒ‰ãƒ»ä¾å­˜é–¢ä¿‚   | chore(deps): Update React to 18.2            |
| perf     | ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹æ”¹å–„ | perf(Button): Reduce bundle size             |
| ci       | CI è¨­å®šå¤‰æ›´        | ci: Update GitHub Actions workflow           |

### 4.4 ã‚³ãƒŸãƒƒãƒˆãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ä½œæˆä¾‹

```bash
# âœ… Good
git commit -m "feat(Button): Add size variant support

Add small, medium, large size options to Button component.

- Update Button.spec.md with size requirements
- Implement size prop in Button.tsx
- Add TC-002 test cases
- Update CSS Modules

Closes #456"

# âŒ Bad
git commit -m "ãƒœã‚¿ãƒ³ã‚’ä¿®æ­£ã—ãŸ"
git commit -m "update button"
git commit -m "WIP"
```

---

## 5. ã‚¨ãƒ©ãƒ¼ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸

### 5.1 ãƒ¦ãƒ¼ã‚¶ãƒ¼å‘ã‘ï¼ˆæ—¥æœ¬èªï¼‰

ã‚¨ãƒ³ãƒ‰ãƒ¦ãƒ¼ã‚¶ãƒ¼ã«è¡¨ç¤ºã•ã‚Œã‚‹ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã¯æ—¥æœ¬èªï¼š

```typescript
if (!isValidEmail(email)) {
  throw new Error("ãƒ¡ãƒ¼ãƒ«ã‚¢ãƒ‰ãƒ¬ã‚¹ã®å½¢å¼ãŒæ­£ã—ãã‚ã‚Šã¾ã›ã‚“ã€‚");
}

showToast("ãƒ•ã‚¡ã‚¤ãƒ«ãŒæ­£å¸¸ã«ä¿å­˜ã•ã‚Œã¾ã—ãŸã€‚");

alert("å…¥åŠ›å†…å®¹ã‚’ç¢ºèªã—ã¦ãã ã•ã„ã€‚");
```

### 5.2 é–‹ç™ºè€…å‘ã‘ï¼ˆè‹±èªï¼‰

ãƒ­ã‚°ã€console.error ã¯è‹±èªï¼š

```typescript
console.error("Invalid email format provided: ", email);
console.warn("Deprecated function used. Use newFunction() instead.");
logger.info("User logged in", { userId, timestamp });
```

### 5.3 ãƒãƒ©ãƒ³ã‚¹ã®å–ã‚Šæ–¹

```typescript
// âœ… Good: ãƒ¦ãƒ¼ã‚¶ãƒ¼ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã¯æ—¥æœ¬èªã€ãƒ­ã‚°ã¯è‹±èª
try {
  await saveData(data);
} catch (error) {
  // ãƒ¦ãƒ¼ã‚¶ãƒ¼ã«è¡¨ç¤º
  showToast("ä¿å­˜ã«å¤±æ•—ã—ã¾ã—ãŸã€‚");

  // é–‹ç™ºè€…å‘ã‘ãƒ­ã‚°
  console.error("Data save failed:", error);
  logger.error("Save operation failed", {
    userId,
    timestamp,
    error: error.message,
  });
}
```

---

## 6. è¤‡æ•°è¨€èªãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆ

### 6.1 å®Ÿè£…ä¾‹

ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã«è‹±èªç‰ˆãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆãŒå¿…è¦ãªå ´åˆï¼š

```
docs/
â”œâ”€â”€ README.md              # æ—¥æœ¬èªï¼ˆãƒ¡ã‚¤ãƒ³ï¼‰
â”œâ”€â”€ README_en.md           # è‹±èªï¼ˆå‚è€ƒï¼‰
â””â”€â”€ rules/
    â”œâ”€â”€ README.md          # æ—¥æœ¬èª
    â”œâ”€â”€ README_en.md       # è‹±èª
```

### 6.2 ä½œæˆãƒ«ãƒ¼ãƒ«

- ãƒ¡ã‚¤ãƒ³è¨€èªï¼ˆæ—¥æœ¬èªï¼‰ã‚’å…ˆã«ä½œæˆ
- è‹±èªç‰ˆã¯å¾Œä»˜ã‘
- ãƒ•ã‚¡ã‚¤ãƒ«åã« `_en` ã‚µãƒ•ã‚£ãƒƒã‚¯ã‚¹

---

## 7. ãƒã‚§ãƒƒã‚¯ãƒªã‚¹ãƒˆ

### ã‚³ãƒ¼ãƒ‰å†…ã‚³ãƒ¡ãƒ³ãƒˆ

- [ ] ã™ã¹ã¦ã®ã‚³ãƒ¡ãƒ³ãƒˆãŒè‹±èªã‹ç¢ºèª
- [ ] ã€Œãªãœã€ã‚’èª¬æ˜ã—ã¦ã„ã‚‹ã‹
- [ ] å¤ã„/å»ƒæ­¢äºˆå®šã‚³ãƒ¡ãƒ³ãƒˆã« FIXME/TODO è¡¨è¨˜ãŒã‚ã‚‹ã‹

### ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆ

- [ ] docs/ é…ä¸‹ãŒæ—¥æœ¬èªã‹ç¢ºèª
- [ ] æŠ€è¡“ç”¨èªãŒçµ±ä¸€ã•ã‚Œã¦ã„ã‚‹ã‹
- [ ] é–¢é€£ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆã¸ã®ãƒªãƒ³ã‚¯ãŒã‚ã‚‹ã‹

### é–¢æ•°åãƒ»å¤‰æ•°å

- [ ] ã™ã¹ã¦ãŒè‹±èªã‹ç¢ºèª
- [ ] ã‚­ãƒ£ãƒ¡ãƒ«ã‚±ãƒ¼ã‚¹ / PascalCase ãŒæ­£ã—ã„ã‹

### ã‚³ãƒŸãƒƒãƒˆ

- [ ] Conventional Commits å½¢å¼ã‹ç¢ºèª
- [ ] æœ¬æ–‡ãŒè‹±èªã‹ç¢ºèª
- [ ] ã‚¹ã‚³ãƒ¼ãƒ—ãŒè¨˜è¿°ã•ã‚Œã¦ã„ã‚‹ã‹

---

## ğŸ”— é–¢é€£ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆ

- [å‘½åè¦å‰‡](./naming-conventions.md)
- [AI ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆé§†å‹•é–‹ç™ºã‚¬ã‚¤ãƒ‰](./ai-documentation.md)
- [ã‚³ãƒ¼ãƒ‰å“è³ªåŸºæº–](./code-quality-standards.md)

---

**æœ€çµ‚æ›´æ–°:** 2025-10-22
**ä½œæˆè€…:** AI (Grok Code Fast 1)
