# MessageConverter è©³ç´°ã‚¬ã‚¤ãƒ‰

MessageConverterã¯ã€Lichtblickæ‹¡å¼µæ©Ÿèƒ½ã‚·ã‚¹ãƒ†ãƒ ã®é‡è¦ãªæ©Ÿèƒ½ã®ä¸€ã¤ã§ã€ãƒ‡ãƒ¼ã‚¿å½¢å¼ã‚’å¤‰æ›ã—ã¦æ—¢å­˜ãƒ‘ãƒãƒ«ã§è¡¨ç¤ºå¯èƒ½ã«ã™ã‚‹ä»•çµ„ã¿ã§ã™ã€‚

## ğŸ“– ç›®æ¬¡

- [æ¦‚è¦](#æ¦‚è¦)
- [åŸºæœ¬çš„ãªä»•çµ„ã¿](#åŸºæœ¬çš„ãªä»•çµ„ã¿)
- [å®Ÿè£…æ–¹æ³•](#å®Ÿè£…æ–¹æ³•)
- [å®Ÿç”¨ä¾‹](#å®Ÿç”¨ä¾‹)
- [åˆ¶é™äº‹é …](#åˆ¶é™äº‹é …)
- [ãƒˆãƒ©ãƒ–ãƒ«ã‚·ãƒ¥ãƒ¼ãƒ†ã‚£ãƒ³ã‚°](#ãƒˆãƒ©ãƒ–ãƒ«ã‚·ãƒ¥ãƒ¼ãƒ†ã‚£ãƒ³ã‚°)

## ğŸ¯ æ¦‚è¦

### MessageConverterã¨ã¯

MessageConverterã¯ã€ã‚«ã‚¹ã‚¿ãƒ ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸å½¢å¼ã‚’æ¨™æº–çš„ãªROS/Lichtblickãƒ¡ãƒƒã‚»ãƒ¼ã‚¸å½¢å¼ã«å¤‰æ›ã™ã‚‹æ©Ÿèƒ½ã§ã™ã€‚ã“ã‚Œã«ã‚ˆã‚Šã€æ—¢å­˜ã®ãƒ‘ãƒãƒ«ï¼ˆPlotã€3Dã€Imageãªã©ï¼‰ã§ã‚«ã‚¹ã‚¿ãƒ ãƒ‡ãƒ¼ã‚¿ã‚’è¡¨ç¤ºã§ãã‚‹ã‚ˆã†ã«ãªã‚Šã¾ã™ã€‚

### åˆ©ç”¨ã‚·ãƒŠãƒªã‚ª

1. **æ—¢å­˜ãƒ‘ãƒãƒ«ã®é–“æ¥çš„æ‹¡å¼µ**

   - ã‚«ã‚¹ã‚¿ãƒ ãƒ‡ãƒ¼ã‚¿ã‚’æ—¢å­˜ãƒ‘ãƒãƒ«ã§è¡¨ç¤º
   - æ—¢å­˜ãƒ‘ãƒãƒ«ã®ã‚³ãƒ¼ãƒ‰ã‚’å¤‰æ›´ã›ãšã«æ©Ÿèƒ½è¿½åŠ 

2. **ãƒ‡ãƒ¼ã‚¿å½¢å¼ã®çµ±ä¸€**

   - ç•°ãªã‚‹ã‚½ãƒ¼ã‚¹ã‹ã‚‰ã®ãƒ‡ãƒ¼ã‚¿ã‚’çµ±ä¸€å½¢å¼ã«å¤‰æ›
   - ãƒ¬ã‚¬ã‚·ãƒ¼ã‚·ã‚¹ãƒ†ãƒ ã¨ã®äº’æ›æ€§ç¢ºä¿

3. **ãƒ—ãƒªã‚»ãƒƒãƒˆæ©Ÿèƒ½ã®å®Ÿç¾**
   - è¨­å®šæƒ…å ±ã‹ã‚‰å®Ÿéš›ã®ãƒ‡ãƒ¼ã‚¿ã‚’ç”Ÿæˆ
   - ãƒ†ãƒ³ãƒ—ãƒ¬ãƒ¼ãƒˆãƒ™ãƒ¼ã‚¹ã®ãƒ‡ãƒ¼ã‚¿ä½œæˆ

## ğŸ”§ åŸºæœ¬çš„ãªä»•çµ„ã¿

### ãƒ‡ãƒ¼ã‚¿ãƒ•ãƒ­ãƒ¼

```mermaid
flowchart LR
    A[ã‚«ã‚¹ã‚¿ãƒ ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸] --> B[MessageConverter]
    B --> C[æ¨™æº–ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸]
    C --> D[æ—¢å­˜ãƒ‘ãƒãƒ«]
    D --> E[è¡¨ç¤º]
```

### ã‚¢ãƒ¼ã‚­ãƒ†ã‚¯ãƒãƒ£

```
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ ãƒ‡ãƒ¼ã‚¿ã‚½ãƒ¼ã‚¹                            â”‚
â”‚ â”œâ”€ ROSãƒˆãƒ”ãƒƒã‚¯                          â”‚
â”‚ â”œâ”€ WebSocket                            â”‚
â”‚ â””â”€ ãã®ä»–ã®ã‚½ãƒ¼ã‚¹                       â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
                 â†“
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ MessagePipeline                         â”‚
â”‚ â”œâ”€ ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸å—ä¿¡                       â”‚
â”‚ â”œâ”€ ã‚¹ã‚­ãƒ¼ãƒæ¤œè¨¼                         â”‚
â”‚ â””â”€ Converterå®Ÿè¡Œ                        â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
                 â†“
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ MessageConverter (æ‹¡å¼µæ©Ÿèƒ½)             â”‚
â”‚ â”œâ”€ å…¥åŠ›ã‚¹ã‚­ãƒ¼ãƒæ¤œè¨¼                     â”‚
â”‚ â”œâ”€ ãƒ‡ãƒ¼ã‚¿å¤‰æ›å‡¦ç†                       â”‚
â”‚ â””â”€ å‡ºåŠ›ã‚¹ã‚­ãƒ¼ãƒç”Ÿæˆ                     â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
                 â†“
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ æ—¢å­˜ãƒ‘ãƒãƒ«                              â”‚
â”‚ â”œâ”€ Plot Panel                           â”‚
â”‚ â”œâ”€ 3D Panel                             â”‚
â”‚ â”œâ”€ Image Panel                          â”‚
â”‚ â””â”€ ãã®ä»–                               â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

## ğŸ’» å®Ÿè£…æ–¹æ³•

### 1. åŸºæœ¬çš„ãªå®Ÿè£…

```typescript
// src/index.ts
import { ExtensionContext } from "@lichtblick/suite";

export function activate(extensionContext: ExtensionContext): void {
  extensionContext.registerMessageConverter({
    fromSchemaName: "custom_msgs/SensorData",
    toSchemaName: "sensor_msgs/Temperature",
    converter: convertSensorData,
  });
}

// å¤‰æ›é–¢æ•°
async function convertSensorData(input: CustomSensorData): Promise<TemperatureMessage> {
  return {
    header: {
      stamp: input.timestamp,
      frame_id: input.sensor_id,
    },
    temperature: input.temp_celsius,
    variance: input.temp_variance || 0,
  };
}

// å‹å®šç¾©
interface CustomSensorData {
  timestamp: { sec: number; nanosec: number };
  sensor_id: string;
  temp_celsius: number;
  temp_variance?: number;
}

interface TemperatureMessage {
  header: {
    stamp: { sec: number; nanosec: number };
    frame_id: string;
  };
  temperature: number;
  variance: number;
}
```

### 2. è¤‡é›‘ãªå¤‰æ›å‡¦ç†

```typescript
// è¤‡æ•°ã®å…¥åŠ›ã‚’çµ±åˆã™ã‚‹ä¾‹
import { ExtensionContext, RegisterMessageConverterArgs } from "@lichtblick/suite";

interface RobotConfig {
  robot_id: string;
  model_type: "industrial" | "mobile" | "humanoid";
  position: { x: number; y: number; z: number };
  orientation: { x: number; y: number; z: number; w: number };
}

interface UrdfMessage {
  data: string; // URDF XML content
}

export function activate(extensionContext: ExtensionContext): void {
  const converter = new RobotConfigConverter();

  extensionContext.registerMessageConverter(converter.getConverterConfig());
}

class RobotConfigConverter {
  private urdfCache = new Map<string, string>();
  private readonly cacheExpiration = 5 * 60 * 1000; // 5åˆ†

  getConverterConfig(): RegisterMessageConverterArgs<RobotConfig> {
    return {
      fromSchemaName: "custom_robot/RobotConfig",
      toSchemaName: "std_msgs/String",
      converter: this.convert.bind(this),
    };
  }

  async convert(config: RobotConfig): Promise<UrdfMessage> {
    try {
      // ã‚­ãƒ£ãƒƒã‚·ãƒ¥ãƒã‚§ãƒƒã‚¯
      const cacheKey = `${config.robot_id}_${config.model_type}`;
      let urdfContent = this.urdfCache.get(cacheKey);

      if (!urdfContent) {
        // URDFã‚³ãƒ³ãƒ†ãƒ³ãƒ„ã‚’ç”Ÿæˆã¾ãŸã¯å–å¾—
        urdfContent = await this.generateUrdfContent(config);
        this.urdfCache.set(cacheKey, urdfContent);

        // ã‚­ãƒ£ãƒƒã‚·ãƒ¥ã®è‡ªå‹•å‰Šé™¤
        setTimeout(() => {
          this.urdfCache.delete(cacheKey);
        }, this.cacheExpiration);
      }

      // ä½ç½®ãƒ»å§¿å‹¢æƒ…å ±ã‚’é©ç”¨
      const processedUrdf = this.applyTransformation(urdfContent, config);

      return { data: processedUrdf };
    } catch (error) {
      console.error("URDF conversion failed:", error);
      // ãƒ•ã‚©ãƒ¼ãƒ«ãƒãƒƒã‚¯: ãƒ‡ãƒ•ã‚©ãƒ«ãƒˆãƒ­ãƒœãƒƒãƒˆãƒ¢ãƒ‡ãƒ«
      return { data: this.getDefaultUrdf() };
    }
  }

  private async generateUrdfContent(config: RobotConfig): Promise<string> {
    // ãƒ—ãƒªã‚»ãƒƒãƒˆãƒ¢ãƒ‡ãƒ«ã‹ã‚‰é¸æŠ
    const modelTemplates = {
      industrial: "path/to/industrial_robot.urdf",
      mobile: "path/to/mobile_robot.urdf",
      humanoid: "path/to/humanoid_robot.urdf",
    };

    const templatePath = modelTemplates[config.model_type];

    // å®Ÿéš›ã®å®Ÿè£…ã§ã¯ã€HTTPãƒªã‚¯ã‚¨ã‚¹ãƒˆã‚„ã‚¢ã‚»ãƒƒãƒˆèª­ã¿è¾¼ã¿
    const response = await fetch(templatePath);
    return await response.text();
  }

  private applyTransformation(urdf: string, config: RobotConfig): string {
    // URDF XMLã«ãƒ•ãƒ¬ãƒ¼ãƒ IDå¤‰æ›ã‚’é©ç”¨
    const frameId = `${config.robot_id}_base_link`;
    return urdf.replace(/base_link/g, frameId);
  }

  private getDefaultUrdf(): string {
    return `<?xml version="1.0"?>
      <robot name="default_robot">
        <link name="base_link">
          <visual>
            <geometry>
              <box size="1 1 1"/>
            </geometry>
          </visual>
        </link>
      </robot>`;
  }
}
```

### 3. éåŒæœŸå‡¦ç†ã¨ã‚¨ãƒ©ãƒ¼ãƒãƒ³ãƒ‰ãƒªãƒ³ã‚°

```typescript
interface WeatherData {
  location: string;
  timestamp: number;
  conditions: string;
}

interface VisualizationMarker {
  header: { frame_id: string; stamp: { sec: number; nanosec: number } };
  text: string;
  position: { x: number; y: number; z: number };
  color: { r: number; g: number; b: number; a: number };
}

async function convertWeatherToMarker(weather: WeatherData): Promise<VisualizationMarker> {
  try {
    // å¤–éƒ¨APIã‹ã‚‰è©³ç´°æƒ…å ±ã‚’å–å¾—ï¼ˆã‚¿ã‚¤ãƒ ã‚¢ã‚¦ãƒˆä»˜ãï¼‰
    const enrichedData = await Promise.race([
      enrichWeatherData(weather),
      new Promise((_, reject) => setTimeout(() => reject(new Error("Timeout")), 5000)),
    ]);

    return {
      header: {
        frame_id: "map",
        stamp: {
          sec: Math.floor(weather.timestamp / 1000),
          nanosec: (weather.timestamp % 1000) * 1000000,
        },
      },
      text: `${weather.location}: ${weather.conditions}`,
      position: await getLocationPosition(weather.location),
      color: getConditionColor(weather.conditions),
    };
  } catch (error) {
    console.warn("Weather enrichment failed, using basic data:", error);

    // ãƒ•ã‚©ãƒ¼ãƒ«ãƒãƒƒã‚¯å‡¦ç†
    return {
      header: {
        frame_id: "map",
        stamp: {
          sec: Math.floor(weather.timestamp / 1000),
          nanosec: (weather.timestamp % 1000) * 1000000,
        },
      },
      text: `${weather.location}: ${weather.conditions}`,
      position: { x: 0, y: 0, z: 0 }, // ãƒ‡ãƒ•ã‚©ãƒ«ãƒˆä½ç½®
      color: { r: 1, g: 1, b: 1, a: 1 }, // ç™½è‰²
    };
  }
}

async function enrichWeatherData(weather: WeatherData): Promise<any> {
  // å¤–éƒ¨APIå‘¼ã³å‡ºã—ã®æ¨¡æ“¬å®Ÿè£…
  const response = await fetch(`/api/weather/${weather.location}`);
  return await response.json();
}

async function getLocationPosition(location: string): Promise<{ x: number; y: number; z: number }> {
  // ã‚¸ã‚ªã‚³ãƒ¼ãƒ‡ã‚£ãƒ³ã‚°APIã®å‘¼ã³å‡ºã—ä¾‹
  // å®Ÿéš›ã®å®Ÿè£…ã§ã¯é©åˆ‡ãªAPIã‚’ä½¿ç”¨
  return { x: 0, y: 0, z: 0 };
}

function getConditionColor(conditions: string): { r: number; g: number; b: number; a: number } {
  const colorMap: Record<string, [number, number, number]> = {
    sunny: [1, 1, 0], // é»„è‰²
    cloudy: [0.7, 0.7, 0.7], // ç°è‰²
    rainy: [0, 0, 1], // é’è‰²
    snowy: [1, 1, 1], // ç™½è‰²
  };

  const [r, g, b] = colorMap[conditions.toLowerCase()] || [1, 0, 0]; // ãƒ‡ãƒ•ã‚©ãƒ«ãƒˆèµ¤è‰²
  return { r, g, b, a: 1 };
}
```

## ğŸ¯ å®Ÿç”¨ä¾‹

### 1. URDF Preset Extensionï¼ˆæ—¢å­˜ã®å®Ÿè£…ä¾‹ï¼‰

```typescript
// ã‚«ã‚¹ã‚¿ãƒ ãƒ­ãƒœãƒƒãƒˆè¨­å®šã‹ã‚‰URDFã‚’ç”Ÿæˆ
export function activate(extensionContext: ExtensionContext): void {
  const converter = new UrdfPresetConverter();
  extensionContext.registerMessageConverter(converter.getConverterConfig());
}

class UrdfPresetConverter {
  getConverterConfig(): RegisterMessageConverterArgs<RobotConfig> {
    return {
      fromSchemaName: "custom_robot/RobotConfig",
      toSchemaName: "std_msgs/String",
      converter: this.convert.bind(this),
    };
  }

  async convert(config: RobotConfig): Promise<{ data: string }> {
    // ãƒ—ãƒªã‚»ãƒƒãƒˆãƒ¢ãƒ‡ãƒ«ã®é¸æŠ
    const modelType = this.determineModelType(config);

    // URDFã‚³ãƒ³ãƒ†ãƒ³ãƒ„ã®èª­ã¿è¾¼ã¿
    const urdfContent = await this.loadModel(modelType);

    // ãƒ•ãƒ¬ãƒ¼ãƒ IDå¤‰æ›ã®é©ç”¨
    const processedUrdf = this.applyFrameTransformations(urdfContent, config.frame_id);

    return { data: processedUrdf };
  }
}
```

### 2. ã‚»ãƒ³ã‚µãƒ¼ãƒ‡ãƒ¼ã‚¿ã®å¯è¦–åŒ–

```typescript
// ã‚«ã‚¹ã‚¿ãƒ ã‚»ãƒ³ã‚µãƒ¼ãƒ‡ãƒ¼ã‚¿ã‚’Plotãƒ‘ãƒãƒ«ç”¨ã«å¤‰æ›
interface CustomSensorReading {
  sensor_id: string;
  readings: Array<{
    type: "temperature" | "humidity" | "pressure";
    value: number;
    unit: string;
  }>;
  timestamp: { sec: number; nanosec: number };
}

async function convertSensorToPlotData(sensor: CustomSensorReading): Promise<any> {
  // è¤‡æ•°ã®ç³»åˆ—ãƒ‡ãƒ¼ã‚¿ã«åˆ†å‰²
  const plotData = sensor.readings.map((reading) => ({
    header: {
      stamp: sensor.timestamp,
      frame_id: sensor.sensor_id,
    },
    data: reading.value,
    label: `${reading.type} (${reading.unit})`,
  }));

  return {
    series: plotData,
    timestamp: sensor.timestamp,
  };
}
```

### 3. ç”»åƒãƒ‡ãƒ¼ã‚¿ã®å¤‰æ›

```typescript
// ã‚«ã‚¹ã‚¿ãƒ ç”»åƒå½¢å¼ã‚’sensor_msgs/Imageã«å¤‰æ›
interface CustomImageData {
  format: "base64" | "url" | "raw";
  data: string | Uint8Array;
  width: number;
  height: number;
  encoding: string;
}

async function convertCustomImage(custom: CustomImageData): Promise<any> {
  let imageData: Uint8Array;

  switch (custom.format) {
    case "base64":
      imageData = new Uint8Array(
        atob(custom.data as string)
          .split("")
          .map((c) => c.charCodeAt(0)),
      );
      break;

    case "url":
      const response = await fetch(custom.data as string);
      const arrayBuffer = await response.arrayBuffer();
      imageData = new Uint8Array(arrayBuffer);
      break;

    case "raw":
      imageData = custom.data as Uint8Array;
      break;
  }

  return {
    header: {
      stamp: { sec: Math.floor(Date.now() / 1000), nanosec: 0 },
      frame_id: "camera",
    },
    height: custom.height,
    width: custom.width,
    encoding: custom.encoding,
    step: custom.width * getBytesPerPixel(custom.encoding),
    data: imageData,
  };
}

function getBytesPerPixel(encoding: string): number {
  const encodingMap: Record<string, number> = {
    rgb8: 3,
    rgba8: 4,
    bgr8: 3,
    bgra8: 4,
    mono8: 1,
    mono16: 2,
  };
  return encodingMap[encoding] || 1;
}
```

## âš ï¸ åˆ¶é™äº‹é …

### æŠ€è¡“çš„åˆ¶é™

#### 1. åŒæœŸå‡¦ç†ã®åˆ¶é™

```typescript
// âŒ åŒæœŸçš„ãªãƒ•ã‚¡ã‚¤ãƒ«èª­ã¿è¾¼ã¿ã¯ä¸å¯
function badConverter(input: any): any {
  const fs = require("fs"); // Node.js APIã¯ä½¿ç”¨ä¸å¯
  return fs.readFileSync("./data.json"); // ã‚¨ãƒ©ãƒ¼
}

// âœ… éåŒæœŸå‡¦ç†ã‚’ä½¿ç”¨
async function goodConverter(input: any): Promise<any> {
  const response = await fetch("/api/data");
  return await response.json();
}
```

#### 2. ãƒªã‚½ãƒ¼ã‚¹ã‚¢ã‚¯ã‚»ã‚¹ã®åˆ¶é™

```typescript
// âŒ ç›´æ¥çš„ãªãƒ•ã‚¡ã‚¤ãƒ«ã‚·ã‚¹ãƒ†ãƒ ã‚¢ã‚¯ã‚»ã‚¹ä¸å¯
const localFile = "./assets/model.urdf"; // ã‚¢ã‚¯ã‚»ã‚¹ä¸å¯

// âœ… HTTPãƒªã‚¯ã‚¨ã‚¹ãƒˆã‚„assetsã‚’ä½¿ç”¨
const modelUrl = "https://example.com/model.urdf";
const response = await fetch(modelUrl);
```

#### 3. ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹åˆ¶é™

```typescript
// âŒ é‡ã„å‡¦ç†ã¯é¿ã‘ã‚‹
async function heavyConverter(input: any): Promise<any> {
  // å·¨å¤§ãªãƒ‡ãƒ¼ã‚¿ã®åŒæœŸå‡¦ç†
  const result = processLargeDataset(input.data); // ãƒ–ãƒ­ãƒƒã‚­ãƒ³ã‚°
  return result;
}

// âœ… éåŒæœŸãƒ»åˆ†å‰²å‡¦ç†
async function optimizedConverter(input: any): Promise<any> {
  // ãƒ¯ãƒ¼ã‚«ãƒ¼ã‚„ãƒãƒ£ãƒ³ã‚¯ã«åˆ†å‰²
  return await processInChunks(input.data);
}
```

### è¨­è¨ˆçš„åˆ¶é™

#### 1. ã‚¹ã‚­ãƒ¼ãƒã®äº’æ›æ€§

- å‡ºåŠ›ã‚¹ã‚­ãƒ¼ãƒã¯æ—¢å­˜ã®ãƒ‘ãƒãƒ«ãŒç†è§£ã§ãã‚‹å½¢å¼ã§ã‚ã‚‹å¿…è¦
- ã‚«ã‚¹ã‚¿ãƒ ã‚¹ã‚­ãƒ¼ãƒã¯ä»–ã®ãƒ‘ãƒãƒ«ã§ã¯ä½¿ç”¨ä¸å¯

#### 2. ã‚¨ãƒ©ãƒ¼ä¼æ’­

- Converterã§ã®ã‚¨ãƒ©ãƒ¼ã¯å…ƒã®ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸å…¨ä½“ã«å½±éŸ¿
- ãƒ•ã‚©ãƒ¼ãƒ«ãƒãƒƒã‚¯å‡¦ç†ã®å®Ÿè£…ãŒé‡è¦

#### 3. é †åºä¿è¨¼

- ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã®é †åºãŒå¤‰æ›´ã•ã‚Œã‚‹å¯èƒ½æ€§
- ã‚¿ã‚¤ãƒ ã‚¹ã‚¿ãƒ³ãƒ—ãƒ™ãƒ¼ã‚¹ã®å‡¦ç†ãŒæ¨å¥¨

## ğŸ”§ ãƒˆãƒ©ãƒ–ãƒ«ã‚·ãƒ¥ãƒ¼ãƒ†ã‚£ãƒ³ã‚°

### 1. å¤‰æ›ãŒå®Ÿè¡Œã•ã‚Œãªã„

**ç—‡çŠ¶**: MessageConverterãŒç™»éŒ²ã•ã‚Œã¦ã„ã‚‹ãŒã€å¤‰æ›ãŒå®Ÿè¡Œã•ã‚Œãªã„

**ç¢ºèªç‚¹**:

```typescript
// 1. ã‚¹ã‚­ãƒ¼ãƒåã®ç¢ºèª
console.log("Registered converter:", {
  from: "custom_robot/RobotConfig", // æ­£ç¢ºãªã‚¹ã‚­ãƒ¼ãƒåã‹ï¼Ÿ
  to: "std_msgs/String",
});

// 2. ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã®å½¢å¼ç¢ºèª
context.onRender = (renderState, done) => {
  console.log("Available topics:", renderState.topics);
  console.log("Messages:", renderState.currentFrame);
  done();
};
```

**è§£æ±ºç­–**:

- ã‚¹ã‚­ãƒ¼ãƒåã®æ­£ç¢ºæ€§ã‚’ç¢ºèª
- å…¥åŠ›ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãŒæœŸå¾…ã—ãŸå½¢å¼ã‹ãƒã‚§ãƒƒã‚¯
- é–‹ç™ºè€…ã‚³ãƒ³ã‚½ãƒ¼ãƒ«ã§ã‚¨ãƒ©ãƒ¼ãƒ­ã‚°ã‚’ç¢ºèª

### 2. å¤‰æ›ã‚¨ãƒ©ãƒ¼

**ç—‡çŠ¶**: å¤‰æ›å‡¦ç†ä¸­ã«ã‚¨ãƒ©ãƒ¼ãŒç™ºç”Ÿ

**ãƒ‡ãƒãƒƒã‚°æ–¹æ³•**:

```typescript
async function debugConverter(input: any): Promise<any> {
  try {
    console.log("Input message:", input);

    // ãƒãƒªãƒ‡ãƒ¼ã‚·ãƒ§ãƒ³
    if (!input || typeof input !== "object") {
      throw new Error("Invalid input format");
    }

    const result = await processMessage(input);
    console.log("Conversion result:", result);

    return result;
  } catch (error) {
    console.error("Conversion error:", error);

    // ãƒ•ã‚©ãƒ¼ãƒ«ãƒãƒƒã‚¯
    return getDefaultMessage();
  }
}
```

### 3. ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹ã®å•é¡Œ

**ç—‡çŠ¶**: å¤‰æ›å‡¦ç†ãŒé…ã„ã€UIãŒãƒ–ãƒ­ãƒƒã‚¯ã•ã‚Œã‚‹

**æœ€é©åŒ–æ‰‹æ³•**:

```typescript
class OptimizedConverter {
  private cache = new Map();
  private processingQueue = new Set();

  async convert(input: any): Promise<any> {
    const key = this.getCacheKey(input);

    // ã‚­ãƒ£ãƒƒã‚·ãƒ¥ãƒã‚§ãƒƒã‚¯
    if (this.cache.has(key)) {
      return this.cache.get(key);
    }

    // é‡è¤‡å‡¦ç†ã®é˜²æ­¢
    if (this.processingQueue.has(key)) {
      await this.waitForProcessing(key);
      return this.cache.get(key);
    }

    // å‡¦ç†å®Ÿè¡Œ
    this.processingQueue.add(key);
    try {
      const result = await this.processMessage(input);
      this.cache.set(key, result);
      return result;
    } finally {
      this.processingQueue.delete(key);
    }
  }
}
```

### 4. ãƒ¡ãƒ¢ãƒªãƒªãƒ¼ã‚¯

**ç—‡çŠ¶**: é•·æ™‚é–“ä½¿ç”¨å¾Œã«ãƒ¡ãƒ¢ãƒªä½¿ç”¨é‡ãŒå¢—åŠ 

**å¯¾ç­–**:

```typescript
class MemoryEfficientConverter {
  private cache = new Map();
  private readonly maxCacheSize = 100;
  private readonly cacheExpiration = 5 * 60 * 1000; // 5åˆ†

  async convert(input: any): Promise<any> {
    // ã‚­ãƒ£ãƒƒã‚·ãƒ¥ã‚µã‚¤ã‚ºåˆ¶é™
    if (this.cache.size >= this.maxCacheSize) {
      this.evictOldestEntries();
    }

    const key = this.getCacheKey(input);
    const cached = this.cache.get(key);

    if (cached && Date.now() - cached.timestamp < this.cacheExpiration) {
      return cached.data;
    }

    const result = await this.processMessage(input);
    this.cache.set(key, {
      data: result,
      timestamp: Date.now(),
    });

    return result;
  }

  private evictOldestEntries(): void {
    const entries = Array.from(this.cache.entries());
    entries.sort((a, b) => a[1].timestamp - b[1].timestamp);

    // å¤ã„ã‚¨ãƒ³ãƒˆãƒªã®å‰Šé™¤
    const toRemove = entries.slice(0, Math.floor(this.maxCacheSize / 2));
    toRemove.forEach(([key]) => this.cache.delete(key));
  }
}
```

## ğŸ”— é–¢é€£ãƒªã‚½ãƒ¼ã‚¹

- [æ‹¡å¼µæ©Ÿèƒ½é–‹ç™ºã‚¬ã‚¤ãƒ‰](./extension-development.md)
- [ãƒ‘ãƒãƒ«æ‹¡å¼µæ€§ã«ã¤ã„ã¦](./panel-extensibility.md)
- [@lichtblick/suite API Reference](https://github.com/Lichtblick-Suite/lichtblick/tree/main/packages/suite)
- [ROS Message Specifications](http://wiki.ros.org/msg)
- [Lichtblick Message Schemas](https://github.com/foxglove/schemas)
